shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1272,65 +1284,220 @@ - (void)updateViewportPadding {
 
 #pragma mark - Keyboard events
 
+typedef NS_ENUM(NSInteger, FlutterKeyboardMode) {
+  FlutterKeyboardModeHidden = 0,
+  FlutterKeyboardModeDocked = 1,
+  FlutterKeyboardModeFloating = 2,
+};
+
+- (void)keyboardWillShowNotification:(NSNotification*)notification {
+  // Immediately prior to a docked keyboard being shown or when a keyboard goes from
+  // undocked/floating to docked, this notification is triggered.
+  [self handleKeyboardNotification:notification notificationName:UIKeyboardWillShowNotification];
+}
+
 - (void)keyboardWillChangeFrame:(NSNotification*)notification {
-  NSDictionary* info = [notification userInfo];
+  // Immediately prior to a change in keyboard frame, this notification is triggered.
+  // There are some cases where UIKeyboardWillShowNotification & UIKeyboardWillHideNotification
+  // do not act as expected and this is used to catch those cases.
+  [self handleKeyboardNotification:notification
+                  notificationName:UIKeyboardWillChangeFrameNotification];
+}
 
-  // Ignore keyboard notifications related to other apps.
-  id isLocal = info[UIKeyboardIsLocalUserInfoKey];
-  if (isLocal && ![isLocal boolValue]) {
-    return;
-  }
+- (void)keyboardWillBeHidden:(NSNotification*)notification {
+  // When keyboard is hidden or undocked, this notification will be triggered.
+  [self handleKeyboardNotification:notification notificationName:UIKeyboardWillHideNotification];
+}
 
-  // Ignore keyboard notifications if engine’s viewController is not current viewController.
-  if ([_engine.get() viewController] != self) {
+- (void)handleKeyboardNotification:(NSNotification*)notification
+                  notificationName:(NSNotificationName)notificationName {
+  BOOL ignoreNotification = [self shouldIgnoreKeyboardNotification:notification
+                                                  notificationName:notificationName];
+  if (ignoreNotification) {
```

Comment:
optional nits:
The name of the method "shouldIgnoreKeyboardNotification" is already self explanatory, I personally would skip the extra local var declaration.
```
if ([self shouldIgnoreKeyboardNotification:notification]) {
```